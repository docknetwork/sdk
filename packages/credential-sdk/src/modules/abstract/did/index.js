export { default as AbstractDIDModule } from './module';
export * from './errors';
