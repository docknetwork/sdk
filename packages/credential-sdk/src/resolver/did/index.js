export { default as DIDKeyResolver } from './did-key-resolver';
export { default as DIDJWKResolver } from './did-jwk-resolver';
export { default as DIDResolver } from './did-resolver';
export { default as UniversalResolver } from './universal-resolver';
