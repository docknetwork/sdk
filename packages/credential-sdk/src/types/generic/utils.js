import TypedBytes from './typed-bytes';
import TypedArray from './typed-array';

export class ByteArray extends TypedBytes {}

export class ArrayOfByteArrays extends TypedArray {
  static Class = TypedBytes;
}
