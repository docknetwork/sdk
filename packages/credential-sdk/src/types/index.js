export * from './blob';
export * from './attest';
export * from './did';
export * from './offchain-signatures';
export * from './payload';
export * from './policy';
export * from './public-keys';
export * from './signatures';
export * from './status-list-credential';
export * from './anchor';
export * from './accumulator';
export * from './trust-registry';
