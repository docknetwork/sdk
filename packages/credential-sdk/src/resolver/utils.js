import { fmtIter } from '../utils/generic';

/**
 * Caches last function result and returns it if function called with the same args again.
 * @param {Function}
 * @returns {Function}
 */
export const cacheLast = (fn) => {
  let lastArgs;
  let lastValue;

  return function cached(...args) {
    if (lastArgs !== void 0) {
      let same = true;
      const length = Math.max(lastArgs.length, args.length);
      for (let i = 0; i < length && same; ++i) {
        same &&= lastArgs[i] === args[i];
      }

      if (same) {
        return lastValue;
      }
    }
    lastValue = fn.apply(this, args);
    lastArgs = args;

    return lastValue;
  };
};

/**
 * Ensures that each item is present in the allowed set and that both sets are valid.
 *
 * @template I
 * @template W
 * @param {Iterable<I | W>} items
 * @param {Iterable<T | W>} allowed
 * @param {W} wildcard
 */
export const ensureItemsAllowed = (items, allowed, wildcard) => {
  const itemsSet = new Set(items);
  const allowedSet = new Set(allowed);
  if (items.size === 0) {
    throw new Error('Empty item set provided');
  }
  if (allowed.size === 0) {
    throw new Error('Empty allowed set provided');
  }

  for (const toCheck of [itemsSet, allowedSet]) {
    if (toCheck.has(wildcard) && toCheck.size > 1) {
      throw new Error(
        `Can't have wildcard mixed with other patterns for ${fmtIter(toCheck)}`,
      );
    }
  }

  if (allowedSet.has(wildcard)) return;

  for (const item of itemsSet) {
    if (!allowedSet.has(item)) {
      throw new Error(
        `Item not found in ${fmtIter(allowedSet)}: \`${item.toString()}\``,
      );
    }
  }
};

export const itemsOrWildcard = (items, wildcard) => {
  for (const item of items) {
    if (item === wildcard) {
      return wildcard;
    }
  }

  return [...new Set(items)];
};
