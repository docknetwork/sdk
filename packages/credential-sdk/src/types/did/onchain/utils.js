/* eslint-disable max-classes-per-file */

import { isHexWithGivenByteSize } from '../../../utils/bytes';
import {
  PublicKey, // eslint-disable-line
} from '../../public-keys';

import { Signature } from "../../signatures"; // eslint-disable-line

import { DockDIDByteSize } from './constants';

/**
 * Check if the given identifier is the hex representation of a Dock DID.
 * @param {string} identifier - The identifier to check.
 * @return {void} Throws exception if invalid identifier
 */
export function validateDockDIDHexIdentifier(identifier) {
  // Byte size of the Dock DID identifier, i.e. the `DockDIDQualifier` is not counted.
  if (!isHexWithGivenByteSize(identifier, DockDIDByteSize)) {
    throw new Error(`DID identifier must be ${DockDIDByteSize} bytes`);
  }
}

/**
 * Check if the given identifier is 32 byte valid SS58 string
 * @param {string} identifier - The identifier to check.
 * @return {void} Throws exception if invalid identifier
 */
export function validateDockDIDSS58Identifier(identifier) {
  // base58-check regex
  const regex = new RegExp(/^[5KL][1-9A-HJ-NP-Za-km-z]{47}$/);
  const matches = regex.exec(identifier);
  if (!matches) {
    throw new Error('The identifier must be 32 bytes and valid SS58 string');
  }
}

/**
 * Returns a `DidKey` as expected by the Substrate node
 * @param {PublicKey} publicKey - The public key for the DID. The Keyring is intentionally avoided here as it may not be
 * accessible always, like in case of hardware wallet
 * @param {VerificationRelationship} verRel
 * @returns {object} - The object has structure and keys with same names as expected by the Substrate node
 */
export function createDidKey(publicKey, verRel) {
  return {
    publicKey: publicKey.toJSON(),
    verRels: verRel.value,
  };
}

/**
 *
 * @param {string|DockDidOrDidMethodKey} did - DID as string or an object
 * @param {number} keyId -
 * @param rawSig
 * @param {Signature} sig
 * @returns {object}
 */
export function createDidSig(did, { keyId }, rawSig) {
  const sig = rawSig.toJSON();

  if (did.isDid) {
    return {
      DidSignature: {
        did: did.asDid,
        keyId,
        sig,
      },
    };
  } else if (did.isDidMethodKey) {
    return {
      DidMethodKeySignature: {
        didMethodKey: did.asDidMethodKey,
        sig,
      },
    };
  } else {
    throw new Error(
      `Incorrect DID passed: \`${did}\`, expected instance of either \`DockDid\` or \`DidMethodKey\``,
    );
  }
}
