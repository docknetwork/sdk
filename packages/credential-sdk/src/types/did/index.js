import { TypedEnum } from '../generic';
import { OffchainDidDetailsValue } from './offchain';
import { StoredOnchainDidDetailsValue } from './onchain';

export * from './onchain';
export * from './offchain';
export * from './document';

export class StoredDidDetails extends TypedEnum {}

export class StoredOnchainDidDetails extends StoredDidDetails {
  static Type = 'onChain';

  static Class = StoredOnchainDidDetailsValue;
}
export class StoredOffchainDidDetails extends StoredDidDetails {
  static Type = 'offChain';

  static Class = OffchainDidDetailsValue;
}

StoredDidDetails.bindVariants(
  StoredOnchainDidDetails,
  StoredOffchainDidDetails,
);
