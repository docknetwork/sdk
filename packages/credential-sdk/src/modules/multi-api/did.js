import { AbstractDIDModule } from '../abstract';
import { DIDDocument, NamespaceDid } from '../../types';
import { injectModuleRouter } from './common';

export default class MultiApiDIDModule extends injectModuleRouter(
  AbstractDIDModule,
) {
  /**
   *
   * @param {DIDDocument} didDocument
   * @param {DidKeypair} didKeypair
   * @returns {Promise<void>}
   */
  async createDocument(didDocument, didKeypair, params) {
    const document = DIDDocument.from(didDocument);

    return await this.moduleById(document.id).createDocument(
      document,
      didKeypair,
      params,
    );
  }

  /**
   *
   * @param {DIDDocument} didDocument
   * @param {DidKeypair} didKeypair
   * @returns {Promise<void>}
   */
  async updateDocument(didDocument, didKeypair, params) {
    const document = DIDDocument.from(didDocument);

    return await this.moduleById(document.id).updateDocument(
      document,
      didKeypair,
      params,
    );
  }

  /**
   *
   * @param {*} did
   * @param {DidKeypair} didKeypair
   * @returns {Promise<void>}
   */
  async removeDocument(did, didKeypair, params) {
    const id = NamespaceDid.from(did);

    return await this.moduleById(id).removeDocument(id, didKeypair, params);
  }

  /**
   *
   * @param {*} did
   * @returns {Promise<DIDDocument>}
   */
  async getDocument(did) {
    const id = NamespaceDid.from(did);

    return await this.moduleById(id).getDocument(id);
  }

  /**
   *
   * @param {DIDDocument} didDocument
   * @param {DidKeypair} didKeypair
   * @returns {Promise<void>}
   */
  async createDocumentTx(didDocument, didKeypair) {
    const document = DIDDocument.from(didDocument);

    return await this.moduleById(document.id).createDocumentTx(
      document,
      didKeypair,
    );
  }

  /**
   *
   * @param {DIDDocument} didDocument
   * @param {DidKeypair} didKeypair
   * @returns {Promise<void>}
   */
  async updateDocumentTx(didDocument, didKeypair) {
    const document = DIDDocument.from(didDocument);

    return await this.moduleById(document.id).updateDocumentTx(
      document,
      didKeypair,
    );
  }

  /**
   *
   * @param {*} did
   * @param {DidKeypair} didKeypair
   * @returns {Promise<void>}
   */
  async removeDocumentTx(did, didKeypair) {
    const id = NamespaceDid.from(did);

    return await this.moduleById(id).removeDocumentTx(id, didKeypair);
  }
}
