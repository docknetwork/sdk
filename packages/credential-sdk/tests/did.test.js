import { randomAsHex, encodeAsSS58 } from "../src/utils";

import { DockDidOrDidMethodKey, DockDid, DidMethodKey } from "../src/types/did";

const hexDid = (did) => DockDid.from(did).toHex();

describe("`DockDidOrDidMethodKey.from`", () => {
  test("`DockDid.from` work for a raw did", () => {
    const result = DockDid.from(
      "5DEHasvC9G3eVF3qCsN2VQvEbHYdQtsv74ozZ1ngQQj39Luk"
    );
    expect(result.did.value).toEqual(
      "0x338afad72761cffcd0619b8e00fe64a32f79463143e0e812a76b1030c593bb4e"
    );

    expect(() =>
      DockDidOrDidMethodKey.from(
        "5DEHasvC9G3eVF3qCsN2VQvEbHYdQtsv74ozZ1ngQQj39Luk"
      )
    ).toThrowErrorMatchingSnapshot();
  });

  test("`DidMethodKey.from` work for a raw did:key", () => {
    const result = DidMethodKey.from(
      "zQ3shokFTS3brHcDQrn82RUDfCZESWL1ZdCEJwekUDPQiYBme"
    );
    expect(result.didMethodKey.secp256k1.value).toEqual(
      "0x03874c15c7fda20e539c6e5ba573c139884c351188799f5458b4b41f7924f235cd"
    );
    expect(result.toString()).toEqual(
      "did:key:zQ3shokFTS3brHcDQrn82RUDfCZESWL1ZdCEJwekUDPQiYBme"
    );

    expect(() =>
      DockDidOrDidMethodKey.from(
        "zQ3shokFTS3brHcDQrn82RUDfCZESWL1ZdCEJwekUDPQiYBme"
      )
    ).toThrowErrorMatchingSnapshot();
  });

  test("`DockDidOrDidMethodKey.from`/`DockDid.from` work for did:dock", () => {
    const classes = [DockDidOrDidMethodKey, DockDid];

    for (const Did of classes) {
      const result = Did.from(
        "did:dock:5DEHasvC9G3eVF3qCsN2VQvEbHYdQtsv74ozZ1ngQQj39Luk"
      );
      expect(result.did.value).toEqual(
        "0x338afad72761cffcd0619b8e00fe64a32f79463143e0e812a76b1030c593bb4e"
      );
    }
  });

  test("`DockDidOrDidMethodKey.from`/`DidMethodKey.from` works for did:key secp256k1", () => {
    const classes = [DockDidOrDidMethodKey, DidMethodKey];

    for (const Did of classes) {
      const result = Did.from(
        "did:key:zQ3shokFTS3brHcDQrn82RUDfCZESWL1ZdCEJwekUDPQiYBme"
      );
      expect(result.didMethodKey.secp256k1.value).toEqual(
        "0x03874c15c7fda20e539c6e5ba573c139884c351188799f5458b4b41f7924f235cd"
      );
      expect(result.toString()).toEqual(
        "did:key:zQ3shokFTS3brHcDQrn82RUDfCZESWL1ZdCEJwekUDPQiYBme"
      );
    }
  });

  test("`DockDidOrDidMethodKey.from`/`DidMethodKey.from` works for did:key ed25519", () => {
    const classes = [DockDidOrDidMethodKey, DidMethodKey];

    for (const Did of classes) {
      const result = Did.from(
        "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8"
      );
      expect(result.didMethodKey.ed25519.value).toEqual(
        "0xd713cb7f8624d8648496e01010f2bd72f0dcbbdecdb7036f38c20475f5f429bf"
      );
      expect(result.toString()).toEqual(
        "did:key:z6MktvqCyLxTsXUH1tUZncNdVeEZ7hNh7npPRbUU27GTrYb8"
      );
    }
  });
});

describe("DID utilities", () => {
  test("On input as 40 byte hex, validateDockDIDIdentifier throws error", () => {
    expect(() => new DockDid(randomAsHex(40))).toThrow(
      /Received `40` items while expected `32` by `DockDidValue`/
    );
  });

  test("On input as 30 byte hex, validateDockDIDIdentifier throws error", () => {
    expect(() => new DockDid(randomAsHex(30))).toThrow(
      /Received `30` items while expected `32` by `DockDidValue`/
    );
  });

  test("On input as 32 byte hex, validateDockDIDIdentifier does not throw error", () => {
    expect(() => new DockDid(randomAsHex(32))).not.toThrow();
  });

  test("On input as 33 byte hex, hexDid throws error", () => {
    const hex = randomAsHex(33);
    expect(() => hexDid(hex)).toThrow(
      /Received `33` items while expected `32` by `DockDidValue`/
    );
  });

  test("On input as 32 byte hex, hexDid returns the input", () => {
    const hex = randomAsHex(32);
    expect(hexDid(hex)).toBe(hex);
  });

  test("On input invalid ss58 but with qualifier, hexDid throws error", () => {
    const did = `${DockDid.Qualifier}oO12`;
    // Without the qualifier, the function tries to parse as hex
    expect(() => hexDid(did)).toThrow(/Invalid ss58/);
  });

  test("On input fully qualified Dock DID, hexDid returns valid hex representation", () => {
    // create a valid DID
    const hex = randomAsHex(32);
    const did = `${DockDid.Qualifier}${encodeAsSS58(hex)}`;
    expect(hexDid(did)).toBe(hex);
  });

  test("On input valid SS58 and with qualifier but smaller than 32 bytes, hexDid throws error", () => {
    const hex = randomAsHex(8);
    const did = `${DockDid.Qualifier}${encodeAsSS58(hex)}`;
    // Without the qualifier, the function tries to parse as hex
    expect(() => hexDid(did)).toThrow(
      /Received `8` items while expected `32` by `DockDidValue`/
    );
  });

  test("On input valid SS58 and with qualifier but larger than 32 bytes, hexDid throws error", () => {
    const ss58 = encodeAsSS58(randomAsHex(32));
    const did = `${DockDid.Qualifier}${ss58}${ss58}`;
    // Without the qualifier, the function tries to parse as hex
    expect(() => hexDid(did)).toThrow(/Invalid ss58/);
  });

  test("On input valid SS58 identifier but smaller than 32 bytes, validateDockDIDSS58Identifier throws error", () => {
    const ss58 = encodeAsSS58(randomAsHex(8));
    expect(() => DockDid.fromQualifiedString(ss58)).toThrow(
      /Received `8` items while expected `32` by `DockDidValue`/
    );
  });

  test("On input valid SS58 identifier but larger than 32 bytes, DockDid.fromQualifiedString throws error", () => {
    const ss58 = encodeAsSS58(randomAsHex(32));
    const did = `${ss58}${ss58}`;
    expect(() => DockDid.fromQualifiedString(did)).toThrow(
      /Invalid ss58 address/
    );
  });

  test("On input valid SS58 identifier, DockDid.fromQualifiedString does not throw error", () => {
    const ss58 = encodeAsSS58(randomAsHex(32));
    expect(() => DockDid.fromQualifiedString(ss58)).not.toThrow();
  });
});
