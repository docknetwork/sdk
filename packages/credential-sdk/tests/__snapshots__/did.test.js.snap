// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`DockDidOrDidMethodKey.from\` \`DidMethodKey.from\` work for a raw did:key 1`] = `"Unsupported qualified string: \`zQ3shokFTS3brHcDQrn82RUDfCZESWL1ZdCEJwekUDPQiYBme\`, expected with either prefix: [did:dock:, did:key:]"`;

exports[`\`DockDidOrDidMethodKey.from\` \`DockDid.from\` work for a raw did 1`] = `"Unsupported qualified string: \`5DEHasvC9G3eVF3qCsN2VQvEbHYdQtsv74ozZ1ngQQj39Luk\`, expected with either prefix: [did:dock:, did:key:]"`;
